### PR TITLE
Cache cleanup and bug fixes

### PR DESCRIPTION
**Pull request by oldalton**
This addresses following issues:
1. Remove expired access tokens
2. Remove invalid refresh tokens
3. Only return accounts with tokens in cache for that client or family id
4. Disassociate application from its family ID after account removal, so that app cannot continue using family refresh token without interactive login first.
5. Query accounts by application's familyId
6. Removed unnecessary cache accessor methods/minor code cleanup

Common core PR: https://github.com/AzureAD/microsoft-authentication-library-common-for-objc/pull/313

### CODE REVIEW

File: MSAL/src/instance/MSALAccountsProvider.m [hunk 1]
```diff
@@ -180,4 +193,25 @@ - (MSALAccount *)accountForUsername:(NSString *)username
     return [msalAccounts allObjects];
 }
 
+- (MSIDAppMetadataCacheItem *)appmetadataItem
```
Comment (unpluggedk) [inline MSAL/src/instance/MSALAccountsProvider.m]: nit: camelCasing? :p

File: MSAL/test/unit/MSALPublicClientApplicationTests.m [hunk 2]
```diff
@@ -1295,6 +1400,46 @@ - (void)testAccountWithHomeAccountId_whenAccountExistsButNotMatching_shouldRetur
...
+    MSIDConfiguration *configuration = [self msalDefaultConfigurationWithAuthority:@"https://login.microsoftonline.com/common"];
+
+    NSError *error = nil;
+    BOOL result = [self.tokenCacheAccessor saveTokensWithConfiguration:configuration
+                                                              response:msidResponse
+                                                               factory:[MSIDAADV2Oauth2Factory new]
+                                                               context:nil
+                                                                 error:&error];
+    XCTAssertTrue(result);
+    XCTAssertNil(error);
```
Comment (unpluggedk) [inline MSAL/test/unit/MSALPublicClientApplicationTests.m]: nit: not sure if we need to test this here, 
1. saveTokens is already tested
2. if anything, wouldn't we better off seeing the cache result? (i.e./ count?)

File: MSAL/test/unit/MSALPublicClientApplicationTests.m [hunk 3]
```diff
@@ -1295,6 +1400,46 @@ - (void)testAccountWithHomeAccountId_whenAccountExistsButNotMatching_shouldRetur
...
+    // Retrieve cache for a different clientId
+    NSArray *override = @[ @{ @"CFBundleURLSchemes" : @[@"msalmyclient"] } ];
+    [MSALTestBundle overrideObject:override forKey:@"CFBundleURLTypes"];
+
+    NSError *appError = nil;
+    __auto_type application = [[MSALPublicClientApplication alloc] initWithClientId:clientId error:&appError];
+    XCTAssertNil(appError);
+    application.tokenCache = self.tokenCacheAccessor;
+
+    NSString *homeAccountId = @"29f3807a-4fb0-42f2-a44a-236aa0cb3f97.0287f963-2d72-4363-9e3a-5705c5b0f031";
```
Comment (unpluggedk) [inline MSAL/test/unit/MSALPublicClientApplicationTests.m]: nit: making it with this guid makes it look like this is a significant value. perhaps change it to generic "non-important-home-accountId" or something?

File: MSAL/test/unit/MSALPublicClientApplicationTests.m [hunk 4]
```diff
@@ -1363,6 +1539,81 @@ - (void)testRemoveAccount_whenAccountExists_shouldRemoveAccount
     XCTAssertEqual([application allAccounts:nil].count, 0);
 }
 
+- (void)testRemoveAccount_whenAccountExists_andIsFociClient_shouldRemoveAccount_andMarkClientNonFoci
```
Comment (unpluggedk) [inline MSAL/test/unit/MSALPublicClientApplicationTests.m]: :+1

File: MSAL/test/unit/MSALPublicClientApplicationTests.m [hunk 5]
```diff
@@ -1363,6 +1539,81 @@ - (void)testRemoveAccount_whenAccountExists_shouldRemoveAccount
...
+
+    NSError *error = nil;
+    BOOL result = [self.tokenCacheAccessor saveTokensWithConfiguration:configuration
+                                                              response:msidResponse
+                                                               factory:[MSIDAADV2Oauth2Factory new]
+                                                               context:nil
+                                                                 error:&error];
+    XCTAssertTrue(result);
+    XCTAssertNil(error);
+
```
Comment (unpluggedk) [inline MSAL/test/unit/MSALPublicClientApplicationTests.m]: nit: count perhaps as that is more significant in testing the rest of the flow?

File: MSAL/test/unit/MSALPublicClientApplicationTests.m [hunk 6]
```diff
@@ -1363,6 +1539,81 @@ - (void)testRemoveAccount_whenAccountExists_shouldRemoveAccount
...
+    XCTAssertEqualObjects([application allAccounts:nil][0], msalAccount);
+
+    // 3. Remove account
+    result = [application removeAccount:msalAccount error:&error];
+
+    XCTAssertTrue(result);
+    XCTAssertNil(error);
+
+    // 4. Make sure the account is now gone
+    XCTAssertEqual([application allAccounts:nil].count, 0);
```
Comment (unpluggedk) [inline MSAL/test/unit/MSALPublicClientApplicationTests.m]: 👍

File: MSAL/test/unit/MSALPublicClientApplicationTests.m [hunk 7]
```diff
@@ -1363,6 +1539,81 @@ - (void)testRemoveAccount_whenAccountExists_shouldRemoveAccount
...
+
+    MSIDRefreshToken *fociToken = [self.tokenCacheAccessor getRefreshTokenWithAccount:account.accountIdentifier familyId:@"1" configuration:configuration context:nil error:nil];
+    XCTAssertNotNil(fociToken);
+
+    MSIDRefreshToken *mrrtToken = [self.tokenCacheAccessor getRefreshTokenWithAccount:account.accountIdentifier familyId:nil configuration:configuration context:nil error:nil];
+    XCTAssertNil(mrrtToken);
+
+    [self msalAddDiscoveryResponse:authorityUrl appendDefaultHeaders:YES];
+
+    // 5. Try to acquire token silently
```
Comment (unpluggedk) [inline MSAL/test/unit/MSALPublicClientApplicationTests.m]: nit: additional comment on what this indicates? i.e./ ... indicating foci is not being used? or something in lines of that

File: MSAL/src/instance/MSALAccountsProvider.m (hunk 1 above)
Comment (oldalton) [reply MSAL/src/instance/MSALAccountsProvider.m]: Fixed

File: MSAL/test/unit/MSALPublicClientApplicationTests.m (hunk 2 above)
Comment (oldalton) [reply MSAL/test/unit/MSALPublicClientApplicationTests.m]: Addressed

File: MSAL/test/unit/MSALPublicClientApplicationTests.m (hunk 3 above)
Comment (oldalton) [reply MSAL/test/unit/MSALPublicClientApplicationTests.m]: Addressed

File: MSAL/test/unit/MSALPublicClientApplicationTests.m (hunk 5 above)
Comment (oldalton) [reply MSAL/test/unit/MSALPublicClientApplicationTests.m]: Addressed

File: MSAL/test/unit/MSALPublicClientApplicationTests.m (hunk 7 above)
Comment (oldalton) [reply MSAL/test/unit/MSALPublicClientApplicationTests.m]: Added a comment

**Pull request merged**